### ISSUE TITLE
Add Ability to Commit - Opt-In

### ISSUE THREAD

**Issue by freimer** [terraform-providers/terraform-provider-panos#6]
As an opt-in, allow for the commit to Panorama and/or Firewalls, and the push of the config from Panorama to the Firewalls.

**Comment by troxil** [comment]
To be honest, this is very dangerous. Even with `panos_security_policies` replacing the entire ruleset of the device - I wouldn't be comfortable even if it was an optional thing..

**Comment by shinmog** [comment]
Terraform itself doesn't have a way for me to define a "finalizer" type of method that I could hook into for commits; that would need to be in place first.  I was going to open a new Terraform issue for that, but it looks like there's this one that's been around for quite a while:

hashicorp/terraform#6258

**Comment by megakoresh** [comment]
This could also be a resource. The users would then have to manually put `depends_on` clause where they define which resources will be commited.

```hcl
resource panos_commit {
  depends_on = [
    "panos_security_policy.myfirstpolicy",
    "panos_security_policy.mysecondpolicy",
  ]
}
```

**Comment by jonshern** [comment]
Yeah commit is definitely a requirement.  
Currently we can call api directly but it is not as clean as doing it all in terraform.
Would be nice to have something all in terraform.

**Comment by erikpaasonen** [comment]
@megakoresh 's solution is essentially what I got working in our POC (`null_provider` resource with a `local_exec` provisioner). I found that the `depends_on` list does not necessarily have to grow and grow, but can simply be replaced every time with just those resources which were changed since the last `terraform apply`. It doesn't have to continue depending on resources which aren't being modified in the current plan.

Sure would be great to have that hook built in to Terraform. Will comment/upvote that Issue. Thanks for the link @shinmog .

**Comment by gliptak** [comment]
partial commit capability was added to ```panos``` https://github.com/PaloAltoNetworks/pango/pull/14

**Comment by shinmog** [comment]
Since HashiCorp has directed provider authors to use the Terraform Plugin SDK, the issue that is blocking native commit support is now located here:

https://github.com/hashicorp/terraform-plugin-sdk/issues/63

**Comment by paultyng** [comment]
I commented on the SDK issue, but wanted to follow up with you all. I think there are maybe a few ways to approach this and also my understanding of how Panorama/Firewalls work could use some clarifying. Specifically, I'm curious what happens when something isn't committed, do subsequent reads of the API get the "dirty" / uncommitted state, or do they get the committed state? Does uncommitted state hang around indefinitely? 

As to workarounds, one method was already suggested above, which is the commit resource with `depends_on`. Another, slightly related method, would be to potentially have resources or data sources output the config/XML and have the commit resource build and commit the entire config/XML at once. This gives you a bit more standard dependency graph. There may be possible other solutions and depending on exactly how state works, we still may be able to utilize the `Close`. The main problems with doing so though are that commit state is not tracked on individual resources, so it wouldn't be east to know if it was committed or not if an error occurred, etc. The fundamental problem though is that this upends the graph, any downstream dependencies (ie. some other resource outside of panos, a DNS record for example) shouldn't be in the graph after the panos resource, but further down the graph after a potential "commit" happens. Terraform is also not transactional, so it would be the responsibility of the provider to handle that, which could be challenging.

**Comment by shinmog** [comment]
@paultyng 

I am in the HashiCorp Terraform Slack channel, so we can discuss more there; but I'll answer the question here so you can refer other other HashiCorp people here instead of sharing Slack DMs.

PAN-OS supports querying both the candidate config (changes made but not yet committed) and the running config (what the firewall is actually doing right now).  The panos provider always reads the candidate config today.  Candidate config hangs around until it's committed or discarded (revert to running config).  Even if PAN-OS allows you to configure something, it may be an invalid config until it is tied to some other piece of config the user has in their plan file.  So a world where a commit happens after every resource is configured is not only a very slow world (commits can take minutes on large config), but also an incorrectly configured world.  So customers will need to be able to make multiple config changes and then commit at the end.

Much to (probably) everyone chagrin, I have thus far resisted making a `panos_commit` that users could then have every resource `depend_on` for numerous reasons:

* Logistically, firewall configs are **huge**.  Hundreds of security rules are commonplace in our bigger customers, and managing security rules compose only 4 of over 120 resources I've developed up until today (and I've still only scratched the surface of what customers need to fully automate).  I cannot imagine asking customer to maintain a `depends_on` clause in a plan file of this magnitude.
* Functionally, I'm not even completely sure how this would work without consistent "configuration drift".  Assume you haven't deployed any config yet, and you have your massive `depends_on`, when you do first deploy, commit happens at the end and you're fine; but what about updates?  You'd have to `terraform apply` twice every time: 1) Read determines that only resource "foo" needs to be updated, so update that  2) Read determines that a commit needs to happen, so commit.  If `panos_commit` could `depend_on` all other resources and I could specify, "and if any of the things you depend on are modified, then you need to run, too", that could work, but that is very close in nature to [another issue](https://github.com/hashicorp/terraform/issues/2341) Terraform has had open for a while.
* Hopefully the most important counterpoint:  a `panos_commit` with a massive `depends_on` would be a hack workaround.

As for using `Close()` to implement commits; I had the same thought process; I'd stuff commits into the `Close()`, then add an opt-in boolean to the provider for commits; but if the commit fails, I need to handle this.  I think that if all HashiCorp did was expose `Close()` and nothing else, I would implement, "commit, and if it fails then discard candidate config" in `Close()`.  The local state would be wrong, but as long as `Close()` returns and displays errors to the user, they would know both that an error happened and what it was.  Furthermore, subsequent `terraform refresh/plan/apply` would do the right thing.  Any user opting-in to commits would know that the local state is wrong tho, and this could just be handled in the documentation.

Tho actually...  discarding the candid config isn't entirely necessary, just tell the user what the error was so they can add / update the plan file accordingly...  Because all the config they've done to get to that point is correct, just not holistically, so re-creating everything is probably too heavy-handed.  ...Yeah, I just need `Close()` to be exposed and able to return errors to the user, and that might be good enough.

If anyone in this thread / watching this issue feels differently, please speak up now :)

I'll hold off on my thoughts of what a Real Solution might look like, as it seems like messing with the Terraform graph is a non-starter.

**Comment by paultyng** [comment]
Here is a PR adding a `Stop` hook to the v2 SDK: https://github.com/hashicorp/terraform-plugin-sdk/pull/377

**Comment by sekka1** [comment]
Hello, any updates to this issue?

**Comment by mlanci** [comment]
Still an issue years later.. Having to create some automation to run firewall-commit -h isn't playing nice.

Is this still being worked on?
